File: cpp_nn/initializers.cpp
#include "initializers.h"


float_batch Constant::initialize(unsigned int w, unsigned int h)
{
    float_batch temp(w, vector<float>(h, 1));

    for(unsigned int i  = 0; i < w; i++){
        for(unsigned int j = 0; j < h; j++){
            temp[i][j] = this->c;
        }
    }
    return temp;
}

float_batch RandomUniform::initialize(unsigned int w, unsigned int h){
    float_batch temp(w, vector<float>(h, 1));
    std::mt19937 gen(this->rd());
    for(unsigned int i  = 0; i < w; i++){
        for(unsigned int j = 0; j < h; j++){
            temp[i][j] = this->dis(gen);
        }
    }
    return temp;
}

float_batch XavierUniform::initialize(unsigned int fan_in, unsigned int fan_out)
{
    float std = sqrt(2 / (fan_in + fan_out));
    float a = std * sqrt(3);

    float_batch temp(fan_in, vector<float>(fan_out, 1));
    std::mt19937 gen(this->rd());
    std::uniform_real_distribution<float> dis{-a, a};
    for(unsigned int i  = 0; i < fan_in; i++){
        for(unsigned int j = 0; j < fan_out; j++){
            temp[i][j] = dis(gen);
        }
    }
    return temp;

}

HeNormal::HeNormal(string non_linearity, string mode)
{
    this->non_linearity = non_linearity;
    this->mode = mode;
    if (this->mode.compare("fan_in") && this->mode.compare("fan_out")){
        throw std::invalid_argument("Invalid mode!");
    }
}

float_batch HeNormal::initialize(unsigned int fan_in, unsigned int fan_out)
{
    int fan = !this->mode.compare("fan_in") ? fan_in : fan_out;
    float gain;
    if (!this->non_linearity.compare("relu")){
        gain = sqrt(2);
    }
    else{
        throw std::runtime_error("invalid nonlinearity");
    }
    float std = gain / sqrt(fan);

    float_batch temp(fan_in, vector<float>(fan_out, 1));
    std::mt19937 gen(this->rd());
    std::normal_distribution<float> dis{0, std};
    for(unsigned int i  = 0; i < fan_in; i++){
        for(unsigned int j = 0; j < fan_out; j++){
            temp[i][j] = dis(gen);
        }
    }
    return temp;
}

File: framework/pure_nn/initializers.py
from abc import ABC
import random
from .activations import Activation, ReLU
import math


def supported_initializers():
    return [x.__name__ for x in Initializer.__subclasses__()]


class Initializer:
    def initialize(self, x):
        raise NotImplementedError


class Constant(Initializer, ABC):
    def __init__(self, c=0):
        self._c = c

    def initialize(self, x):
        if isinstance(x[0], int):
            w = len(x)
            temp = [None for _ in range(w)]
            for i in range(w):
                temp[i] = self._c
            return temp
        elif isinstance(x[0], list):
            w, h = len(x), len(x[0])
            temp = [[None for _ in range(h)] for _ in range(w)]
            for i in range(w):
                for j in range(h):
                    temp[i][j] = self._c
            return temp
        else:
            raise TypeError


class RandomUniform(Initializer, ABC):
    def initialize(self, x):
        w, h = len(x), len(x[0])
        temp = [[None for _ in range(h)] for _ in range(w)]
        for i in range(w):
            for j in range(h):
                temp[i][j] = random.uniform(0, 1)
        return temp


class XavierUniform(Initializer, ABC):
    def initialize(self, x):
        fan_in, fan_out = len(x), len(x[0])  # TODO: only works for Dense layer!
        std = math.sqrt(2 / (fan_in + fan_out))
        a = std * math.sqrt(3)

        temp = [[None for _ in range(fan_out)] for _ in range(fan_in)]
        for i in range(fan_in):
            for j in range(fan_out):
                temp[i][j] = random.uniform(-a, a)
        return temp


class HeNormal(Initializer, ABC):
    def __init__(self, non_linearity, mode="fan_in"):
        if not isinstance(non_linearity, Activation):
            raise Exception()
        self.non_linearity = non_linearity
        self.mode = mode

    def initialize(self, x):
        fan_in, fan_out = len(x), len(x[0])  # TODO: only works for Dense layer!
        fan = fan_in if self.mode == "fan_in" else fan_out
        if isinstance(self.non_linearity, ReLU):
            gain = math.sqrt(2)
        else:
            raise NotImplementedError
        std = gain / math.sqrt(fan)

        temp = [[None for _ in range(fan_out)] for _ in range(fan_in)]
        for i in range(fan_in):
            for j in range(fan_out):
                temp[i][j] = random.gauss(0, std)
        return temp
